Document template-let deprecation (#417)
digital-asset/docs.daml.com: docs/2.8.0/docs/daml/reference/templates.rst

.. _daml-ref-template-let:

> Template-local Definitions (Deprecated)

Reviewer: Hi @akrmn.  Can we add a before and after code example to make this really concrete?  I think this one will do https://daholdings.slack.com/archives/C01UC8LP76Y/p1692122312487989?thread_ts=1692115771.111369&cid=C01UC8LP76Y but leave it up to you.

Author: thank you @hrischuk-da that's a great example, I'll add it shortly

Author: done now

   - interface methods applied to ``this``

   in a template-local let definition will result in an error generated by the infinite loop.

.. warning::

>    Since Daml 2.8.0, template-local definitions are deprecated and their

Author: I added this in the docs for 2.8.0 given that 2.7.0 is about to be released. Perhaps 2.7.1 would make more sense but we don't have a folder for that yet.

Reviewer 2: 2.7.0 is not out yet, if this is something you and the team feel strongly about there is still time to put it in the release notes and add that one line to the docs.

Reviewer 2: (We definitely don't deprecate things in patch releases, so it's either 2.7.0 or 2.8.0.)

   - interface methods applied to ``this``

   in a template-local let definition will result in an error generated by the infinite loop.

> .. warning::

Reviewer: Can we remove the 'warning' label about the infinite loop discussion?  Otherwise, there are two warnings in a row and that means people will miss the importance of the second warning.  I also think the infinite loop can be used as justification for deprecating the feature.

Author: done, I moved the infinite loop discussion into the deprecation warning as justification, thanks for the suggestion. Could you have another look at that please?

   Since Daml 2.8.0, template-local definitions are deprecated and their
   presence will result in the following warning:

   .. code-block:: text

     Template-local binding syntax ("template-let") is deprecated,
     it will be removed in a future version of Daml.
     Instead, use plain top level definitions, taking parameters
     for the contract fields or body ("this") if necessary.

> Migrating

Reviewer 3: ```suggestion
Migration
```

=========

> Users are strongly encouraged to rewrite their code to avoid this feature. Such

Reviewer: Can we change 'rewrite' to 'adapt'?  It is a little softer in tone.

Author: done, thank you

> Users are strongly encouraged to rewrite their code to avoid this feature. Such
> a rewrite would involve replacing each template-local definition with a regular
> top level definition. If the old definition made use of contract fields or the

Reviewer 3: ```suggestion
Users are strongly encouraged to rewrite their code to avoid this feature. This involves replacing each template-local definition with a regular
top-level definition. If the old definition made use of contract fields or the
```

> contract body ("this"), those won't be in scope at the top level, so the new

Reviewer 3: ```suggestion
contract body ("this"), the new
```

definition should take them as parameters. Correspondingly, the use sites of
these definitions should supply the appropriate values as arguments.

> Disabling the warning
> =====================

Reviewer 3: ```suggestion
Deactivate the warning
======================
```

This warning is controlled by the warning flag ``template-let``, which means

> that it can be disabled independently of other warnings. This is especially

Reviewer 3: ```suggestion
that it can be deactivated independently of other warnings. This is especially
```

useful for gradually migrating code that used this syntax.

> To disable the warning within a Daml file, add the following line at the top of

Reviewer 3: ```suggestion
To deactivate the warning within a Daml file, add the following line at the top of
```

the file:

.. code-block:: daml

  {-# OPTIONS_GHC -Wno-template-let #-}

> To disable it for an entire Daml project, add the following entry to the

Reviewer 3: ```suggestion
To deactivate it for an entire Daml project, add the following entry to the
```

``build-options`` field of the project's ``daml.yaml`` file

.. code-block:: yaml

  build-options:
  - --ghc-option=-Wno-template-let

> Within a project where the warning has been disabled as described above, it can
> be reenabled for a single Daml file by adding the following line at the top of

Reviewer 3: ```suggestion
Within a project where the warning has been deactivated for the whole project as described above, it can
be reactivated for a single Daml file by adding the following line at the top of
```
